File: src/main/java/carelog/carelog/common/config/audit/AuditorAwareImpl.java
package carelog.carelog.config.audit;

import org.springframework.data.domain.AuditorAware;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;

import java.util.Optional;

public class AuditorAwareImpl implements AuditorAware<String> {

    private static final String SYSTEM = "SYSTEM";

    @Override
    public Optional<String> getCurrentAuditor() {
        try {
            Authentication authentication = SecurityContextHolder.getContext().getAuthentication();

            if (authentication == null || !authentication.isAuthenticated()) {
                return Optional.of(SYSTEM);
            }

            Object principal = authentication.getPrincipal();

            if (principal instanceof UserDetails userDetails) {
                String username = userDetails.getUsername();
                return Optional.ofNullable(username)
                        .filter(s -> !s.isBlank())
                        .or(() -> Optional.of(SYSTEM));
            }

            if (principal instanceof String s) {
                if (s.isBlank() || "anonymousUser".equalsIgnoreCase(s)) {
                    return Optional.of(SYSTEM);
                }
                return Optional.of(s);
            }
        } catch (Exception ignored) {
            // Fall through to default SYSTEM user
        }

        return Optional.of(SYSTEM);
    }
}

File: src/main/kotlin/carelog/carelog/common/config/audit/AuditorAwareImpl.kt
package carelog.carelog.common.config.audit

import org.springframework.data.domain.AuditorAware
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.core.userdetails.UserDetails
import java.util.Optional

class AuditorAwareImpl : AuditorAware<String> {

    companion object {
        private const val SYSTEM = "SYSTEM"
    }

    override fun getCurrentAuditor(): Optional<String> {
        return try {
            val authentication = SecurityContextHolder.getContext().authentication

            if (authentication == null || !authentication.isAuthenticated) {
                return Optional.of(SYSTEM)
            }

            when (val principal = authentication.principal) {
                is UserDetails -> Optional.ofNullable(principal.username)
                    .filter { it.isNotBlank() }
                    .or { Optional.of(SYSTEM) }
                is String -> if (principal.isBlank() || principal.equals("anonymousUser", ignoreCase = true))
                    Optional.of(SYSTEM)
                else
                    Optional.of(principal)
                else -> Optional.of(SYSTEM)
            }
        } catch (e: Exception) {
            Optional.of(SYSTEM)
        }
    }
}